repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -3,117 +3,153 @@
-import org.springframework.jdbc.core.namedparam.SqlParameterSource;
-import org.springframework.jdbc.support.GeneratedKeyHolder;
-import org.springframework.jdbc.support.KeyHolder;
 import org.springframework.stereotype.Repository;
 
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.criteria.*;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Objects;
 import java.util.Optional;
 
-import static com.epam.esm.repository.Parameters.*;
-import static com.epam.esm.repository.builder.CertificateQueries.*;
+import static com.epam.esm.repository.builder.CertificateQueries.DELETE_TAG_FROM_CERTIFICATE;
+import static com.epam.esm.repository.builder.CertificateQueries.INSERT_TAG_TO_CERTIFICATE;
 
 @Repository
 public class CertificateRepositoryImpl implements CertificateRepository {
 
-    private final CertificateMapper rowMapper;
-    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
-
-    @Autowired
-    public CertificateRepositoryImpl(CertificateMapper rowMapper, NamedParameterJdbcTemplate jdbcTemplate) {
-        this.rowMapper = rowMapper;
-        this.namedParameterJdbcTemplate = jdbcTemplate;
+    private static final String TAGS = "tags";
+    private static final String ID = "id";
+    private static final String NAME = "name";
+    private static final String DESCRIPTION = "description";
+    private static final String PARAMETER = "parameter";
+    private static final String PRICE = "price";
+    private static final String DURATION = "duration";
+    private static final String CREATE_DATE = "createDate";
+    private static final String LAST_UPDATE_DATE = "lastUpdateDate";
+    private static final String PARTIAL_STRING = "%%%s%%";
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    public CertificateRepositoryImpl(EntityManager entityManager) {
+        this.entityManager = entityManager;
     }
 
     @Override
     public List<Certificate> find(SearchCriteria searchCriteria) {
-        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Certificate> query = criteriaBuilder.createQuery(Certificate.class);
+        Root<Certificate> certificateRoot = query.from(Certificate.class);
+        List<Predicate> predicateList = new ArrayList<>();
+        Join<Object, Object> join = certificateRoot.join(TAGS, JoinType.LEFT);
+        query = query.select(certificateRoot);
+
+        //Condition by certificate name
+        if (StringUtils.isNotEmpty(searchCriteria.getCertificateName())) {
+            String partialName = String.format(PARTIAL_STRING, searchCriteria.getCertificateName());
+            Predicate predicate = criteriaBuilder.like(certificateRoot.get(NAME), partialName);
+            predicateList.add(predicate);
+        }

```

Comment:
Please follow the DRY principle

Reply:
Should I use private methods in Repository classes?

Reply:
fixed